openedx/core/djangoapps/util/row_delete.py

```diff
@@ -0,0 +1,101 @@
+"""
+Code to delete rows from a table within a Django mgmt command using best practices.
+"""
+
+import logging
+import time
+from django.core.management.base import CommandError, BaseCommand
+from django.db import transaction
+
+log = logging.getLogger(__name__)
+
+
+def delete_rows(
+        model_mgr,
+        table_name,
+        chunk_size,
+        sleep_between
+    ):
+    """
+    Deletes *ALL* rows from table, chunking the deletes to avoid long table/row locks.
+
+    Args:
+        model_mgr (django.db.models.manager.Manager): Django ORM mgr for the table's model.
+        table_name (str): Name of table from which to delete all rows.
+        chunk_size (int): Number of rows to delete in each transaction.
+        sleep_between (float): Number of seconds to sleep between transactions.
+    """
+    if chunk_size <= 0:
+        raise CommandError('Only positive chunk size is allowed ({}).'.format(chunk_size))
+    if sleep_between < 0:
+        raise CommandError('Only non-negative sleep between seconds is allowed ({}).'.format(sleep_between))
+
+    # The "as id" below fools Django raw query into thinking the primary key is being queried.
+    min_max_ids = model_mgr.raw(
+        'SELECT MIN(id) as id, MAX(id) as max_id FROM {}'.format(table_name)
+    )[0]
+    min_id = min_max_ids.id
+    max_id = min_max_ids.max_id
```

Comment:
Again, a nit, but going strictly by id might create problems where tables are sparse in some ranges of ids, but dense in others. I know we have some tables with 1 row per every 1000 or so ids in the low range, but are 1:1 toward the current max id. The performance characteristics and db load would vary pretty wildly in that scenario and might cause an outage if a large chunk_size was chosen without taking that into account. I've got some ideas if you want to address the problem, but this is probably fine for the vast majority of cases.

Reply:
Understood - that's why I've put the `around` caveats in the log statements. I'm hoping that the runner of this script will understand that ranges of IDs might be missing. I'll add a warning comment to this effect.